ReactNativeClient/lib/services/searchengine/filterParser.ts

```
@@ -0,0 +1,101 @@
+				terms.push([currentCol, currentTerm]);
+				currentTerm = '';
+				inQuote = false;
+			} else {
+				inQuote = true;
+			}
+			continue;
+		}
+
+		if (c === ' ' && !inQuote) {
+			if (!currentTerm) continue;
+			terms.push([currentCol, currentTerm]);
+			currentCol = '_';
+			currentTerm = '';
+			continue;
+		}
+
+		if (c === ':' && !inQuote) {
+			currentCol = currentTerm;
+			currentTerm = '';
+			continue;
+		}
+
+		currentTerm += c;
+	}
+	if (currentTerm) terms.push([currentCol, currentTerm]);
+
+
+	let relation = 'AND'; // default relation
+	const result: Array<Filter> = [];
+	for (let i = 0; i < terms.length; i++) {
+		if (terms[i][0] !== '_') {
+			// Hopefully a valid filter
+			// TO DO : handle negation
+			result.push({ relation: relation, name: terms[i][0], value: terms[i][1] });
+			relation = 'AND'; // reset to default
+		} else {
+			// could be AND or OR or text to fts search
+			if (terms[i][1].toUpperCase() !== 'AND' && terms[i][1].toUpperCase() !== 'OR') {
+				// this is text
+				result.push({ relation: relation, name: 'text', value: terms[i][1] });
+				relation = 'AND'; // reset to default
+			} else {
+				// this is a relation for the next term;
+				relation = terms[i][1].toUpperCase(); // OR
+			}
+		}
+	}
+
+	return result;
+};
+
+const trimQuotes = (str: string): string => str.startsWith('"') ? str.substr(1, str.length - 2) : str;
+
+export function filterParser(searchString: string) {
+	searchString = searchString.trim();
+	const filters: Map<string, Array<Term>> = new Map();
+
+	const matches = parseQuery(searchString);
+	let terms: Array<Term>;
+	for (const match of matches) {
+		const { name, relation, value } = match;
+
+		if (!value) continue;
+
+		if (name === 'title' || name === 'body') {
+			// Trim quotes since we don't support phrase query here
+			// eg. Split title:"hello world" to title:hello title:world with relation
+			terms = trimQuotes(match.value).split(' ').map(value => ({ relation, value }));
```

That should be done in the parser I think. Basically we should expect that what we get from the parser is valid and can be used right away (or if it's invalid the parser should throw an error).